src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/validation/FlowValidationFsm.java

```diff
@@ -116,47 +115,58 @@ protected void receiveData(State from, State to,
             return;
         }
 
-        List<SwitchId> switchIds = service.getSwitchIdListByFlowId(flowId);
+        switchIdSet.addAll(service.getSwitchIdListByFlowId(flowId));
 
-        awaitingRules = switchIds.size();
+        awaitingRules = switchIdSet.size();
         log.debug("Send commands to get rules on the switches");
-        switchIds.forEach(switchId ->
+        switchIdSet.forEach(switchId ->
                 getCarrier().sendSpeakerRequest(flowId, new DumpRulesForFlowHsRequest(switchId)));
 
-        log.debug("Send commands to get meters on the termination switches");
-        awaitingMeters = TERMINATION_SWITCHES_COUNT;
-        getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getSrcSwitchId()));
-        getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getDestSwitchId()));
-
-        log.debug("Send commands to get groups on the termination switches");
-        awaitingGroups = TERMINATION_SWITCHES_COUNT;
-        getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getSrcSwitchId()));
-        getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getDestSwitchId()));
+        awaitingMeters = 1;
+        awaitingGroups = 1;
+        if (!flow.isOneSwitchFlow()) { // avoid duplicates for single sw rules
+            awaitingMeters++;
+            log.debug("Send commands to get meters on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getSrcSwitchId()));
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getDestSwitchId()));
+            awaitingGroups++;
+            log.debug("Send commands to get groups on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getSrcSwitchId()));
+            getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getDestSwitchId()));
+        } else {
+            log.debug("Send commands to get meters on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getSrcSwitchId()));
+            log.debug("Send commands to get groups on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getSrcSwitchId()));
+        }
```

Comment:
looks like a lot of code duplication. You found very elegant solution for rules `awaitingRules = switchIdSet.size();`

I think you can do the same for groups and meters:
```
        Set<SwitchId> terminationSwitchIds = Sets.newHashSet(flow.getSrcSwitchId(), flow.getDestSwitchId());
        awaitingMeters = terminationSwitchIds.size();
        awaitingGroups = terminationSwitchIds.size();

        for (SwitchId terminationSwitchId : terminationSwitchIds) {
            log.debug("Send command to get meters on the termination switch {}", terminationSwitchId);
            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(terminationSwitchId));
            log.debug("Send command to get groups on the termination switch {}", terminationSwitchId);
            getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(terminationSwitchId));
        }
```

Reply:
thank you, this code looks way more elegant